src/GraphODataTemplateWriter/CodeHelpers/Java/TypeHelperJava.cs

```diff
@@ -157,7 +158,61 @@ public static string TemplateName(this string templateFile)
+        public static string CanonicalPath(this OdcmObject c, OdcmModel m)
+        {
+            var propertyInformation = GetCanonicalNavigationProperties(c, m, true); //singleton type all have ContainsTarget=True to support navigation properties bindings, but they don't actually contain target
+            var shortestCanonicalPath = propertyInformation
+                                        .Select(x => GetNavigationPathFromProperty(x, m))
+                                        .Select(x => new { path = x, slashCount = x.Count(y => y == '/') })
+                                        .OrderBy(x => x.slashCount)
+                                        .Select(x => x.path)
+                                        .FirstOrDefault();
+            return string.IsNullOrEmpty(shortestCanonicalPath) ? string.Empty : lastIdReplacer.Replace(shortestCanonicalPath, string.Empty);
+        }
+        public static bool ShouldIncludeReferenceUpdateMethods(this OdcmObject c, OdcmModel m)
+        {
+            return GetCanonicalNavigationProperties(c, m).Any();
+        }
+        /// <summary>
+        /// Enumerates all navigation properties for a specific type in order to determine the canonical path for reference requests
+        /// </summary>
+        /// <param name="c">Type to find navigation properties for</param>
+        /// <param name="m">Model to find navigation properties from</param>
+        /// <param name="includeContainsTarget">Whether or not to include navigation properties that include target.</param>
+        /// <param name="includeNonCollectionProperties">Whether or not to include navigation properties that return single items.</param>
+        /// <returns>navigation properties for a specific type</returns>
+        private static IEnumerable<OdcmProperty> GetCanonicalNavigationProperties(this OdcmObject c, OdcmModel m, bool includeContainsTarget = false, bool includeNonCollectionProperties = false)
+        {
+            Func<OdcmProperty, bool> propertiesFilter = x => x.IsNavigation() && x.Projection.Type == c && (x.IsCollection || includeNonCollectionProperties) && (x.IsReference() || includeContainsTarget);
+            var entityContainerProperties = m.EntityContainer.Properties.Where(propertiesFilter);
+            var namespaceProperties = m.Namespaces
+                                            .SelectMany(x => x.Classes)
+                                            .SelectMany(x => x.Properties.Where(propertiesFilter));
+            return m.EntityContainer.Properties.OfType<OdcmSingleton>()
+                                        .SelectMany(x => x.Projection.Type.AsOdcmClass().Properties.Where(propertiesFilter))
+                                        .Union(entityContainerProperties.Any() ? entityContainerProperties : new List<OdcmProperty>())
+                                        .Union(namespaceProperties.Any() ? namespaceProperties : new List<OdcmProperty>());
```

Comment:
not blocking, but I wonder what you think about this in terms of reading the logic faster:
```suggestion
            return m.EntityContainer.Properties
                .Union(m.Namespaces.SelectMany(x => x.Classes)
                    .SelectMany(x => x.Properties))
                .Union(m.EntityContainer.Properties.OfType<OdcmSingleton>()
                    .SelectMany(x => x.Projection.Type.AsOdcmClass().Properties))
                .Where(propertiesFilter);
```

Reply:
thanks for the suggestion. After going back and forth I forgot to take a step back on this one.
I just pushed a new commit improving upon your suggestion, let me know what you think